feat: integration spring data jdbc docs (#8582)
ydb-platform/ydb: ydb/docs/en/core/integrations/spring-data-jdbc.md

# {{ ydb-short-name }} Dialect for Spring Data JDBC

This guide is intended for use with [Spring Data JDBC](https://spring.io/projects/spring-data-jdbc) and {{ ydb-short-name }}.

Spring Data JDBC is part of the [Spring Data](https://spring.io/projects/spring-data) ecosystem, providing a simplified way to interact with relational databases using SQL and Java objects. Unlike Spring JPA, which relies on JPA (Java Persistence API), Spring Data offers a direct approach to working with databases, bypassing complex ORM (Object-Relational Mapping) for simpler methods.

> ## Installing the {{ ydb-short-name }} Dialect {#install-dialect}
> 
> To integrate {{ ydb-short-name }} with your Spring Data JDBC project, you will need to add two dependencies: the {{ ydb-short-name }} JDBC Driver and the Spring Data JDBC extension for {{ ydb-short-name }}.
> 
> Examples for different build systems:

Reviewer: ```suggestion
## Installing the {{ ydb-short-name }} dialect {#install-dialect}

To integrate {{ ydb-short-name }} with a Spring Data JDBC project, it needs two dependencies: the [{{ ydb-short-name }} JDBC driver](https://github.com/ydb-platform/ydb-jdbc-driver/) and the Spring Data JDBC extension for {{ ydb-short-name }}.

Examples for different build systems:
```

{% list tabs %}

- Maven

    ```xml
    <!-- Set actual versions -->
    <dependency>
        <groupId>tech.ydb.jdbc</groupId>
        <artifactId>ydb-jdbc-driver</artifactId>
        <version>${ydb.jdbc.version}</version>
    </dependency>

    <dependency>
        <groupId>tech.ydb.dialects</groupId>
        <artifactId>spring-data-jdbc-ydb</artifactId>
        <version>${spring.data.jdbc.ydb}</version> 
    </dependency>
    ```

- Gradle

    ```groovy
    dependencies {
        // Set actual versions
        implementation "tech.ydb.dialects:spring-data-jdbc-ydb:$ydbDialectVersion"
        implementation "tech.ydb.jdbc:ydb-jdbc-driver:$ydbJdbcVersion"
    }
    ```

{% endlist %}

## Usage {#using}

> After importing all the necessary dependencies, the dialect is ready for use. Let's consider a simple example of a Spring Data JDBC application.

Reviewer: ```suggestion
After importing all the necessary dependencies, the dialect is ready for use. Below is a simple example of a Spring Data JDBC application.
```

```properties

spring.datasource.driver-class-name=tech.ydb.jdbc.YdbDriver
spring.datasource.url=jdbc:ydb:<grpc/grpcs>://<host>:<2135/2136>/path/to/database[?saFile=file:~/sa_key.json]
```

```java

@Table(name = "Users")
public class User implements Persistable<Long> {
    @Id
    private Long id = ThreadLocalRandom.current().nextLong();
    
    private String login;
    private String firstname;
    private String lastname;
    
    @Transient
    private boolean isNew;
    
    // Constructors, getters and setters
    
    @Override
    public Long getId() {
      return id;
    }
    
    @Override
    public boolean isNew() {
      return isNew;
    }
    
    public void setNew(boolean isNew) {
      this.isNew = isNew;
    }
}
```

> For the `User` entity in the `Users` table, let's create a repository:

Reviewer: ```suggestion
Creating a repository for the `User` entity in the `Users` table:
```

```java

public interface SimpleUserRepository extends CrudRepository<User, Long> {
}
```

> Let's save a new user and verify that they have been successfully saved:

Reviewer: ```suggestion
Saving a new user and verifying that it has been successfully saved:
```

```java

@Component
public class UserRepositoryCommandLineRunner implements CommandLineRunner {
    
    @Autowired
    private SimpleUserRepository repository;
    
    @Override
    public void run(String... args) {
        User user = new User();
        user.setLogin("johndoe");
        user.setFirstname("John");
        user.setLastname("Doe");
        user.setNew(true);  // Setting the flag for the new entity
      
        // Save user
        User savedUser = repository.save(user);
      
        // Check saved user
        assertThat(repository.findById(savedUser.getId())).contains(savedUser);
      
        System.out.println("User saved with ID: " + savedUser.getId());
    }
}
```

### View Index {#viewIndex}

> To generate `VIEW INDEX` statements from repository methods, you need to use the `@ViewIndex` annotation. The `@ViewIndex` annotation has two fields:
> - `indexName` - the name of the index.
> - `tableName` - the name of the table to which `VIEW INDEX` is bound, especially useful when using the @MappedCollection annotations.
> 
> Let's consider a simple example with an index on the Users table by the `login` field:

Reviewer: ```suggestion
To generate `VIEW INDEX` statements from repository methods, use the `@ViewIndex` annotation. The `@ViewIndex` annotation has two fields:

- `indexName`: The index name.
- `tableName`: The table name to which the `VIEW INDEX` is bound, which is particularly useful when using the `@MappedCollection` annotation.

Here is an example of an index on the Users table by the `login` field:
```

```Java
public interface SimpleUserRepository extends CrudRepository<User, Long> {

    @ViewIndex(indexName = "login_index")
    User findByLogin(String login);
}
```

The query generated by this method will look as follows:

```sql
SELECT 
    `Users`.`id` AS `id`, 
    `Users`.`login` AS `login`, 
    `Users`.`lastname` AS `lastname`, 
    `Users`.`firstname` AS `firstname` 
FROM `Users` VIEW login_index AS `Users` WHERE `Users`.`login` = ?
```

### YdbType {#ydbType}

> To declare a specific data type in {{ ydb-short-name }} you can use the `@YdbType` annotation on the entity field. Example usage:

Reviewer: ```suggestion
The `@YdbType` annotation allows you to declare a specific {{ ydb-short-name }} data type for an entity field. Here is an example of its usage:
```

```kotlin
@YdbType("Json")
lateinit var jsonColumn: String
@YdbType("JsonDocument")
lateinit var jsonDocumentColumn: String
@YdbType("Uint8")
lateinit var uint8Column: Byte
@YdbType("Uint16")
lateinit var uint16Column: Short
@YdbType("Uint32")
lateinit var uint32Column: Int
@YdbType("Uint64")
lateinit var uint64Column: Long
```

> Using the `@YdbType` annotation allows you to accurately specify data types supported by {{ ydb-short-name }}, ensuring proper interaction with the database.
> 
> An example of a simple Spring Data JDBC repository can be found at the following [link](https://github.com/ydb-platform/ydb-java-examples/tree/master/jdbc/spring-data-jdbc).

Reviewer: ```suggestion
Using the `@YdbType` annotation allows you to accurately specify the data types supported by {{ ydb-short-name }}, ensuring proper interaction with the database.

A complete example of a simple Spring Data JDBC repository is available [on GitHub](https://github.com/ydb-platform/ydb-java-examples/tree/master/jdbc/spring-data-jdbc).
```
